### PR TITLE
Add positive_nonlinearity option to likelihoods and kernels

### PR DESCRIPTION
**Pull request by jacobrgardner**
This PR adds support for using transformations other than `torch.exp` to ensure positiveness of hyperparameters by passing in a different nonlinearity to most built in modules that have parameters that must be positive. For example:

```python
softplus = torch.functional.F.softplus

class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(ExactGPModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.RBFKernel(positive_nonlinearity=softplus),
            positive_nonlinearity=softplus
        )
    
    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

likelihood = GaussianLikelihood(positive_nonlinearity=sofplus)
model = ExactGPModel(train_x, train_y, likelihood)
```

It is likely in the future that `torch.functional.F.softplus` will become the default choice in the near future because it circumvents issues that arise when large steps during optimization lead to large parameter value changes.

### CODE REVIEW

**Comment by jacobrgardner** [comment]
@bletham Using softplus instead of exp partially solves #370 in the sense that Adam in your notebook now works up to learning rates of 6. 

@samuelstanton Using softplus may actually make training substantially more stable for your application as well, since you deal with particularly extreme hyperparameter values.

**Comment by jacobrgardner** [comment]
I've been playing around with using `softplus` instead of `exp` in some of our VI notebooks where we've previously had some trouble with stability issues, and the results are pretty amazing actually. 

In the SVDKL regression examples, replacing `exp` with `softplus` and no other change results in a 10% relative decrease in mean absolute error, and makes convergence much less sensitive to the exact learning rate chosen.

Let's give it some time to play around with the option and make sure we understand all of the ramifications, but I'm definitely in favor of sunsetting use of `exp` as the default option here ASAP.

**Comment by jacobrgardner** [comment]
Regarding dropping `log`: ideally yes, we would drop these. There are concerns about this being breaking, however (e.g. saved models would no longer load properly). 

We could do that change now; however my preference is to change the names once we are confident in using softplus as the default option so that all of the breaking changes for that happen at once.

**Comment by Balandat** [comment]
Seems reasonable. What about the prior kwargs? Those shouldn't affect model loading if the priors themselves are still stored with the `log_` prefix. But I guess that's not really worth the effort...

**Comment by jacobrgardner** [comment]
I think ideally we rename `log_*_prior` to `*_prior`, and the `transform` (we can change the name to this, agreed) gets passed along to the prior.

**Comment by gpleiss** [comment]
@Balandat @jacobrgardner - since this is sort of a beta feature, I'm okay with keeping the `log_*` names for right now. We can add another issue/PR for doing a soft deprecation of the names.

**Comment by gpleiss** [comment]
I'm okay with merging this if @Balandat is!
